### DISCUSSION TITLE
Help in faster navigation

### DISCUSSION THREAD

**Discussion by hasantahir** [godbout/Wooshy.docs#91] [Q&A]
Greetings,

As the screengrab explains, is there any way to reach the intended target faster? For example, in Chrome (or any other web browser), Wooshy searches for tabs and then the active page. Any way to bypass the tabs and focus on the current page?

Thanks.

https://user-images.githubusercontent.com/11236148/208137836-5592ca33-90aa-43f3-8888-821adc25ab96.mov

**Comment by godbout** [comment]
hello Hasan!

i had the idea before of building some kind of "priority" levels, so that people reach stuff before others. also like for example the current app rather than the menu bar of the status items first. but i'm not sure it's gonna be very useful, because priorities may change for the same user depending on the situation. does that make sense?

for your current case i would use the `role` filter: https://github.com/godbout/Wooshy.docs#search
i would probable type `log but` which would filter the `login` `button`. (or whatever other role it is. depends on how the web page was built. could be a `link` so in that case `log l` may work instead.)

hope this helps! lemme know.

**Reply by hasantahir** [reply]
Thanks for your response as always G.

In the settings, there is an option to configure the scope. Although I love to access UI elements through Wooshy in apps such as Word, Outlook, Powerpoint etc., navigating in a browser becomes a bit inefficient (personal take). If my memory is serving me right, one of the earlier versions of Readme pointed to the roles like `log but` etc, but not anymore.

As asked earlier as well, at least in a web browser, I would love the scope to be on the page. Also curious about what is the fuzzy search about. Does it learn from experience?

cheers

**Reply by godbout** [reply]
> In the settings, there is an option to configure the scope. Although I love to access UI elements through Wooshy in apps such as Word, Outlook, Powerpoint etc., navigating in a browser becomes a bit inefficient (personal take). If my memory is serving me right, one of the earlier versions of Readme pointed to the roles like `log but` etc, but not anymore.

i'm not a big fan of inconsistency. making Wooshy behave differently for webpages would be a bit unsettling, i think ☹️

that's possible that there was a `log but`. the text has been changed, but the explanation should still be there somehow (point 2 in link about the README above, filtering by role). there's also now two short videos, the second one showing the use of roles.

maybe it's not clear enough. i'm marking this down to be reviewed. thanks. 

> As asked earlier as well, at least in a web browser, I would love the scope to be on the page. 

so, for a webpage, Wooshy would ignore all the rest? Dock, App Menu, Status Items, even if they're part of the scope? (and even ignoring tabs etc. it seems, in the description you're giving.)

> Also curious about what is the fuzzy search about. Does it learn from experience?

fuzzy search is not about learning from experience. on the same web page, the same search will always return the same results, in the same order. for something like Alfred, learning from experience makes sense, because you have one set of data (what's on the computer). for Wooshy that would mean a set of data for each app, webpage, etc. 

what fuzzy search means is that you don't need to type the whole words in order. for example if you want to target "hello my friend how are you?" you can just type "he fri".

**Reply by godbout** [reply]
@hasantahir any chance to precise what is not clear in the README regarding the Search, for you? thanks.

edit: i've written down also about your idea of tweaking the Targets specifically for web pages/web browsers.

**Reply by godbout** [reply]
i've made some changes in the README but ultimately i'm not sure whether it's better 🤣️

**Reply by godbout** [reply]
@hasantahir regarding your original question, i don't think changing the order of the Targets is great, because then to navigate them it's gonna be tricky. which one is first? which one is last?
so what about there's new options in Wooshy where, for web browsers, you can decide to ignore the URL bar, and the main tabs? would that help?

**Reply by godbout** [reply]
@hasantahir also, with the crazy number of tabs you're having, how do you get to another tab? do you go back/forward one by one? or do you use any other extension that allows you to go directly to one, in the same way Wooshy would allow you to by typing parts of its title?

**Comment by hasantahir** [comment]
Hi @godbout 

Happy new year!

If README mentions the different kinds of attributes that Wooshy can access, that would be great.

For example, there are `button`, `link`, `pop-up button`, `tab`, `bookmark button`, `text entry area` etc. that I have learned through the Inspector. I think Wooshy is already smart enough to shorten the aforementioned types. I also think putting them in the README would help in personalising the app's use, as different persons can shorten the elements differently and settle on something.

**Reply by godbout** [reply]
Happy New Year my friend!

> If README mentions the different kinds of attributes that Wooshy can access, that would be great.

what do you mean by `Attributes`? what i call `Attributes` personally are all the things Wooshy grabs **in the background**. so it grabs not only what you see, but all the metadata behind. but i believe that bothering users with that info is not useful. i wanna keep the README as simple and straightforward as possible. no extra info. then users can learn how to get comfortable with Wooshy by using it. but maybe you're talking about something else?

> For example, there are `button`, `link`, `pop-up button`, `tab`, `bookmark button`, `text entry area` etc. that I have learned through the Inspector. I think Wooshy is already smart enough to shorten the aforementioned types. I also think putting them in the README would help in personalising the app's use, as different persons can shorten the elements differently and settle on something.

hmm. again, i do still currently find the README too complicated, and messy. not sure yet how to make it simpler tho, but i believe i'll find a way. but listing all the UI Elements roles feels wrong to me. what feels right is letting users know about The Inspector, so they can build their own patterns themselves. some may like using roles, some don't. but listing the roles, hmm, i can't see how this helps. new roles may be added in the future by Apple also, or changed (some changed between macOS 12 and macOS 13).

the idea of Wooshy—but maybe it doesn't work for everyone—was just to type what comes to your mind: like what you see, and/or the type of element. like you try, out of the blue. wanna reach the close button? you try "clo butt". most of the time it should work. but when it doesn't, you can use The Inspector to see what could reach what you wanna reach. of course that is helpful only if you deal with apps or webpages that you use regularly.

i agree that the README has to be better, but listing all the roles or attributes, i see that as a downgrade, and as too much information for the user.

what _may_ work though is having a first, gross, global README, that then point to more other detailed ones? but again, i think if a user sees a list of roles and feels like they need to remember them, that's gonna give a bad taste. better use Wooshy, understand the principle, and use The Inspector when needed. no?

**Reply by godbout** [reply]
regarding your original question about faster navigation, i think it would be great if you open an issue about this. i think it's a good idea to add an option in Wooshy to target differently in webpages/webbrowsers, as you are suggesting.

**Reply by godbout** [reply]
oh also not sure if that'll help but i have some new videos for the YT Channel regarding The Inspector (and other things).

**Reply by hasantahir** [reply]
Thanks for sharing the Youtube channel. Watching at 0.25X to figure out the keyboard moves :-). Would be nice to have enlarged key casting for future videos. On keeping the README simple, I agree. However, not sure about the philosophy of the Inspector. It feels like an additional learning item to get the roles of the elements. It has immense value for troubleshooting purposes, but it would be inefficient to call it every time something new pops up.

Cheers.

**Reply by godbout** [reply]
The Inspector is exactly for what you're talking about: debugging. you should try to reach an Element with natural language. if you can't find how to reach an Element, then The Inspector is here to show you what can be used to target that Element. but you shouldn't have to use if all the time, or even often. it should just be there as an help, if you don't find how to reach an Element.

the world of AX is complex, and the user shouldn't bother with that at all. it doesn't make sense to ask user to remember every technical AX UI Element role and subroles that exist in macOS (and that Apple changes from time to time). just try to use common sense. see a button? type button. see a tab? type tab. if can't find how to reach, check out The Inspector.
